### PR TITLE
US77132 - View updates UI portion

### PR DESCRIPTION
**Pull request by ctwomey1**
UI work for US77132.  

can be tested by manually using setCourseUpdates method in course tile.

### CODE REVIEW

File: d2l-course-tile.html [hunk 1]
```diff
@@ -67,9 +67,15 @@
...
-					<template id="courseCodeTemplate" is="dom-if" if="[[showCourseCode]]">
-						<span class="course-code-text">{{_organization.properties.code}}</span>
+				<div class="course-info">
+					<div class="course-text d2l-heading-4">{{_organization.properties.name}}
+
+						<template id="courseCodeTemplate" is="dom-if" if="[[showCourseCode]]">
+							<span class="course-code-text">{{_organization.properties.code}}</span>
+						</template>
+					</div>
+					<template id="courseUpdates" is="dom-if" if="{{_hasCourseUpdates}}">
```
Comment (ctwomey1) [inline d2l-course-tile.html]: I know there's a bit of perf contention using dom if's for small things.  However for rendering i want the box to not be present so that the word wrapping works correctly for course name.  Totally up for suggestions of making that still work with the hidden$= attribute.

Comment (apalaniuk) [inline d2l-course-tile.html]: I think we can just `display: none` it - the `hidden` attribute can be tricky based on other applied CSS IIRC

File: d2l-course-tile-styles.html [hunk 2]
```diff
@@ -131,6 +131,27 @@
...
+		}
+		.update-text-box {
+			border: 2px solid #E57231;
+			color: #E57231;
+			border-radius: 0.4rem;
+			font-size: 16px;
+			width: 2rem;
+			height: 2rem;
+			display: inline-block;
+			padding-left: 8px;
```
Comment (jstefaniuk-d2l) [inline d2l-course-tile-styles.html]: Use shorthand for padding:
/* padding: top right bottom left */

File: d2l-course-tile-styles.html [hunk 3]
```diff
@@ -131,6 +131,27 @@
...
+			flex: 1;
+		}
+		.course-info {
+			display: flex;
+		}
+		.update-text-box {
+			border: 2px solid #E57231;
+			color: #E57231;
+			border-radius: 0.4rem;
+			font-size: 16px;
```
Comment (jstefaniuk-d2l) [inline d2l-course-tile-styles.html]: this should be translated to `0.8rem` in case the user changes the root font size

File: d2l-course-tile-styles.html [hunk 4]
```diff
@@ -131,6 +131,27 @@
...
 			margin-top: 0.4rem;
+			float: left;
+			flex: 1;
+		}
+		.course-info {
+			display: flex;
+		}
+		.update-text-box {
+			border: 2px solid #E57231;
+			color: #E57231;
```
Comment (jstefaniuk-d2l) [inline d2l-course-tile-styles.html]: Does this map to a vui color? http://design.d2l/fundamentals/colour/

File: d2l-course-tile.html [hunk 5]
```diff
@@ -294,6 +303,21 @@
 						break;
 				}
 			},
+			setCourseUpdates: function(updates) {
+				if (updates > 99) {
+					this._courseUpdates = '99+';
+					this._hasCourseUpdates = true;
+				} else if (updates > 0 && updates <= 99) {
+					this._courseUpdates = updates.toString();
```
Comment (jstefaniuk-d2l) [inline d2l-course-tile.html]: My guess is that you can do here without `toString`

File: d2l-course-tile-styles.html [hunk 6]
```diff
@@ -131,6 +131,27 @@
 			width: 100%;
```
Comment (jstefaniuk-d2l) [inline d2l-course-tile-styles.html]: I think you can remove this since it's overrided by the flex property

File: d2l-course-tile-styles.html (hunk 4 above)
Comment (apalaniuk) [inline d2l-course-tile-styles.html]: `d2l-color-carnelian`

**Comment by ctwomey1** [comment]
If someone could go through and test each browser for ui deficiencies that'd be great.  Make sure to go through the 3 scenarios of no updates, 1 digit updates, 2 digit updates, 99+ updates

Testing completed:
- [ ] ie
- [ ] firefox
- [ ] safari
- [ ] chrome

File: d2l-course-tile-styles.html [hunk 7]
```diff
@@ -17,6 +17,9 @@
 		.d2l-course-tile-hidden {
 			display: none;
 		}
+		.d2l-updates-hidden {
```
Comment (jstefaniuk-d2l) [inline d2l-course-tile-styles.html]: You can probably just use `.d2l-course-tile-hidden` for this one, think of the title for the course-tile hidden as more of a namespacing thing than an actual description thing (otherwise we'd have a ton of hidden classes with the exact same thing in them)

(Disclaimer: This is just an opinion)

Comment (ctwomey1) [inline d2l-course-tile-styles.html]: it's valid, I created a new one because the other was so specific.  Realistically we could create our own name spaced class d2l-hidden that lives in a more central area instead of making a new one each page. For now i'll leave it as is :man_dancing:

File: d2l-course-tile-styles.html [hunk 8]
```diff
@@ -128,9 +131,25 @@
...
+			border: 2px solid var(--d2l-color-carnelian);
+			color: var(--d2l-color-carnelian);
+			border-radius: 0.4rem;
+			font-size: 0.8rem;
+			width: 2.1rem;
+			height: 2.1rem;
+			line-height: 1.9rem;
+			display: inline-block;
+			margin-top: 0.4rem;
+   			text-align: center;
```
Comment (jstefaniuk-d2l) [inline d2l-course-tile-styles.html]: (nit) spacing

File: d2l-course-tile-styles.html (hunk 7 above)
Comment (ryantmer) [inline d2l-course-tile-styles.html]: Yarr, `d2l-course-tile-hidden` was just kind of my overreaction to an issue that came up with using `hidden` as a class. Probably went a bit far in the other direction; oh well!

File: test/d2l-course-tile/d2l-course-tile.js [hunk 9]
```diff
@@ -464,4 +464,30 @@ describe('<d2l-course-tile>', function() {
 			expect(widget.$.telemetryRequest.generateRequest.called).to.equal(true);
 		});
 	});
+
+	describe('setting course updates attribute', function() {
+		it('should show update number when less than 99', function() {
```
Comment (CodeBaboon) [inline test/d2l-course-tile/d2l-course-tile.js]: Sort of dumb but a test for `99` proper might be useful (to make sure you don't accidentally get '99+').

File: test/d2l-course-tile/d2l-course-tile.js [hunk 10]
```diff
@@ -464,4 +464,30 @@ describe('<d2l-course-tile>', function() {
...
+		it('should show 99+ when more than 99 updates', function() {
+			widget.setCourseUpdates(100);
+			expect(widget.$.courseUpdates.getAttribute('class')).to.not.contain('d2l-updates-hidden');
+			expect(widget._courseUpdates).to.equal('99+');
+			expect(widget._hasCourseUpdates).to.be.true;
+			expect(widget.$$('.update-text-box').innerText).to.equal('99+');
+
+		});
+
+		it('should not show updates number when 0', function() {
```
Comment (CodeBaboon) [inline test/d2l-course-tile/d2l-course-tile.js]: What if updates is < 0, null, or undefined?

File: d2l-course-tile.html [hunk 11]
```diff
@@ -67,10 +67,16 @@
...
 					</div>
 				</div>
-				<div class="course-text d2l-heading-4">{{_organization.properties.name}}
-					<template id="courseCodeTemplate" is="dom-if" if="[[showCourseCode]]">
-						<span class="course-code-text">{{_organization.properties.code}}</span>
-					</template>
+				<div class="course-info">
+					<div class="course-text d2l-heading-4">{{_organization.properties.name}}
+
+						<template id="courseCodeTemplate" is="dom-if" if="[[showCourseCode]]">
```
Comment (CodeBaboon) [inline d2l-course-tile.html]: Based on @ryantmer's perf findings this should maybe get switched from a `dom-if` to CSS hidden while we are in here.

File: d2l-course-tile.html [hunk 12]
```diff
@@ -294,6 +302,21 @@
 						break;
 				}
 			},
+			setCourseUpdates: function(updates) {
```
Comment (CodeBaboon) [inline d2l-course-tile.html]: Nothing calls this but tests right now. What will call this?

File: d2l-course-tile.html (hunk 11 above)
Comment (ryantmer) [inline d2l-course-tile.html]: > _kneels down_
> _puts ear to ground_
> _listens carefully_
> _raises head_

..._Si, por favor mi amigo._

Comment (jstefaniuk-d2l) [inline d2l-course-tile.html]: 🚂

File: d2l-course-tile.html (hunk 12 above)
Comment (ctwomey1) [inline d2l-course-tile.html]: see code review #270 Basically originally i thought i had to do more work in the lms... ultimately i split these code reviews in 2 for little good reason...

File: d2l-course-tile.html (hunk 11 above)
Comment (ctwomey1) [inline d2l-course-tile.html]: I feel like this should be a separate pull request.  Mostly cause i want to close off this unit of work and it is technically unrelated to this change

Comment (ryantmer) [inline d2l-course-tile.html]: Yeah it's not a huge performance problem or anything. Just as long as it doesn't get forgotten; there are a few cases where it can probably be changed, all at once, as a low-priority sort of thing.

Comment (CodeBaboon) [inline d2l-course-tile.html]: Yeah I was thinking more along the whole boyscout motto lines but I'm fine with doing it in a different PR.

File: d2l-course-tile-styles.html [hunk 13]
```diff
@@ -128,9 +131,25 @@
 			text-transform: uppercase;
 		}
 		.course-text {
-			width: 100%;
 			position: relative;
 			margin-top: 0.4rem;
+			float: left;
```
Comment (apalaniuk) [inline d2l-course-tile-styles.html]: I think `space-between` on the parent flexbox may be cleaner, but it seems minor and there might be other styling that I'm not seeing.

File: d2l-course-tile.html [hunk 14]
```diff
@@ -73,11 +73,17 @@
...
-						<span class="course-code-text">{{_organization.properties.code}}</span>
-					</template>
-					<span aria-label$="[[_courseNotificationLabel]]"></span>
+
+				<div class="course-info">
+					<div class="course-text d2l-heading-4">{{_organization.properties.name}}
+						<template id="courseCodeTemplate" is="dom-if" if="[[showCourseCode]]">
+							<span class="course-code-text">{{_organization.properties.code}}</span>
+						</template>
+            <span aria-label$="[[_courseNotificationLabel]]"></span>
```
Comment (apalaniuk) [inline d2l-course-tile.html]: nit: spacing

**Comment by trevorjonesd2l** [comment]
Could we make this more clear for screen readers? Right now when I navigate with NVDA on Firefox it will read:
>"{Course name} {Number of updates}"
> "20th Century French Theatre 99"

Probably need Bob's opinion on this but maybe it could be along the lines of

> "20th Century French Theatre 99 updates" or "20th Century French Theatre with 99 updates"

**Comment by trevorjonesd2l** [comment]
**Tests**

Chrome
- [X] 99+ number
- [X] 99+ with different zoom levels
- [X] Double digit number
- [X] Single digit number
- [X] No updates

Firefox
- [X] 99+ number
- [X] 99+ with different zoom levels
- [X] Double digit number
- [X] Single digit number
- [X] No updates
- [X] Screen Reader

Internet Explorer
- [X] 99+ number
- [X] 99+ with different zoom levels
- [X] Double digit number
- [X] Single digit number
- [X] No updates

Edge
- [X] 99+ number
- [X] 99+ with different zoom levels
- [X] Double digit number
- [X] Single digit number
- [X] No updates

Safari
- [ ] 99+ number
- [ ] 99+ with different zoom levels
- [ ] Double digit number
- [ ] Single digit number
- [ ] No updates

Chrome Mobile iPhone 6
- [X] 99+ number
- [X] Double digit number
- [X] Single digit number
- [X] No updates
- [X] Scrolling

Chrome RTL
- [X] 99+ number
- [X] Double digit number
- [X] Single digit number
- [X] No updates

**Comment by trevorjonesd2l** [comment]
Some issues I found while testing. I consider the first two changes necessary to fix. The last one (IE) would be nice to have.

**1 - Updates on screen reader**

Could we make this more clear for screen readers? Right now when I navigate with NVDA on Firefox it will read:

> "{Course name} {Number of updates}"
> "20th Century French Theatre 99"

Probably need Bob's opinion on this but maybe it could be along the lines of

> "20th Century French Theatre 99 updates" or "20th Century French Theatre with 99 updates"

**2 - Overflow on mobile with long course names**

image


**3 - Vertical alignment on IE**

The update text doesn't seem to be aligned in the center like the rest of the browsers (IE is dumb)
image

**Comment by ctwomey1** [comment]
pushed my fix for the first two issues, i'll look into the ie thing further

**Comment by ctwomey1** [comment]
ie fix and slight modification to localization based on conversation with bob/christian

**Comment by trevorjonesd2l** [comment]
All issues have been addressed. I'll verify it behaves correctly in Safari tomorrow
